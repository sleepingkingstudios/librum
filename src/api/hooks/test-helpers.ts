import type {
  Alert,
  AlertsContext,
} from '@alerts';
import type { Dispatch } from '@store';
import type { EffectOptions } from '../types';
import type {
  Response,
  UseQueryResult,
} from './types';

const alerts: AlertsContext = {
  alerts: [] as Alert[],
  dismissAlert: jest.fn(),
  dismissAllAlerts: jest.fn(),
  displayAlert: jest.fn(),
}
const dispatch: Dispatch = jest.fn();

export const defaultOptions: EffectOptions = {
  alerts,
  dispatch,
};

export const defaultResponse: Response = {
  hasData: false,
  hasError: false,
  isUninitialized: false,
  isLoading: false,
  isErrored: false,
  isFailure: false,
  isSuccess: false,
  status: 'unknown',
};

export const defaultResult: UseQueryResult = {
  isUninitialized: false,
  isLoading: false,
  isSuccess: false,
  isError: false,
  refetch: jest.fn(),
};
