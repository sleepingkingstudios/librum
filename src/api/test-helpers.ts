import { defaultTo } from 'lodash';

import type { DataObject } from '@utils/types';
import type {
  ApiError,
  FetchFailure,
  FetchSuccess,
} from './types';

interface IBuildFailureResponseOptions {
  data?: DataObject;
  error?: ApiError;
  status?: number;
}

interface IBuildSuccessResponseOptions {
  data?: DataObject;
}

export const buildFailureResponse = (
  options?: IBuildFailureResponseOptions
): FetchFailure => {
  if (options === null || options == undefined) {
    return { error: { data: {}, status: 400 }};
  }

  const data = defaultTo(options.data, null);
  const error = defaultTo(options.error, null);
  const status = defaultTo(options.status, 400);

  if (error === null) {
    return { error: { data: {}, status }};
  }

  if (data === null) {
    return {
      error: {
        status,
        data: {
          ok: false,
          error,
        },
      },
    };
  }

  return {
    error: {
      status,
      data: {
        ok: false,
        data,
        error,
      },
    },
  };
};

export const buildSuccessResponse = (
  options?: IBuildSuccessResponseOptions
): FetchSuccess => {
  if (options === null || options == undefined) {
    return { data: { ok: true } }
  }

  const data = defaultTo(options.data, null);

  if (data === null) {
    return { data: { ok: true } }
  }

  return { data: { data, ok: true } };
};

export const fetchErrorResponse: FetchFailure = {
  error: {
    error: 'something went wrong',
    status: 'FETCH_ERROR',
  },
};

export const serializedErrorResponse: FetchFailure = {
  error: { message: 'something went wrong' }
};
