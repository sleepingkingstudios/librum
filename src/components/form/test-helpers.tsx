import * as React from 'react';
import {
  Form,
  Formik,
  FormikValues,
} from 'formik';

import type { OnSubmit } from './types';

interface IFormWrapperProps {
  initialValues?: FormikValues;
  onSubmit?: OnSubmit;
  submitButton?: true | false;
}

const SubmitButton = (
  { submitButton }: { submitButton: true | false }
): JSX.Element => {
  if (!submitButton) { return null; }

  return (
    <button type="submit">Submit</button>
  );
};

export const formWrapper = (
  { initialValues, onSubmit, submitButton }: IFormWrapperProps) => {
  const FormWrapper = (
    { children }: { children: React.ReactNode }
  ): JSX.Element => (
    <Formik initialValues={initialValues || {}} onSubmit={onSubmit}>
      <Form>
        { children }

        <SubmitButton submitButton={submitButton} />
      </Form>
    </Formik>
  );

  return FormWrapper;
};
