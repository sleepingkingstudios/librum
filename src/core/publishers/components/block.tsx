import * as React from 'react';
import { Link } from 'react-router-dom';
import { FontAwesomeIcon as Icon } from '@fortawesome/react-fontawesome';
import { faLeftLong } from '@fortawesome/free-solid-svg-icons';

import { DataList } from '@components/data-list';
import type { DataListData } from '@components/data-list';
import { ExternalLink } from '@components/external-link';

type MaybePublisher = {
  name?: string,
  slug?: string,
  website?: string,
};

const defaultPublisher: MaybePublisher = {
  name: null,
  slug: null,
  website: null,
};

const formatData = (publisher: MaybePublisher): DataListData => {
  const { name, slug, website } = publisher;

  return {
    name: name || ' ',
    slug: slug || ' ',
    website: website ? (<WebsiteLink to={website.toString()} />) : ' ',
  };
};

const WebsiteLink = ({ to }: { to: string }): JSX.Element => (
  <ExternalLink to={to} />
);

export const PublisherBlock = ({
  data,
}: {
  data: Record<string, Record<string, unknown>>,
}): JSX.Element => {
  const publisher =
    'publisher' in data ? data.publisher as MaybePublisher : defaultPublisher;
  const formatted = formatData(publisher);

  return(
    <>
      <DataList data={formatted} />

      <Link to="/publishers">
        <Icon className="mr-1" icon={faLeftLong} />

        Back to Publishers
      </Link>
    </>
  );
};
