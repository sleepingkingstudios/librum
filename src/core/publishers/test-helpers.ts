import type { Publisher } from './types';

export const publishers: Publisher[] = [
  {
    id: 'e7c781df-ed94-45a5-90ee-94d28cc8f2dc',
    name: 'Crystal Sphere Publications',
    slug: 'crystal-sphere-publications',
    website: 'www.example.com/crystal-sphere-publications',
  },
  {
    id: 'eb79903b-3b73-420e-8e5a-604f27138260',
    name: 'Flumph Free Press',
    slug: 'flumph-free-press',
    website: 'www.example.com/flumph-free-press',
  },
  {
    id: 'dc40f9c7-69e5-47b5-bcde-d062bc6c44b1',
    name: 'Spelljammer Monthly',
    slug: 'spelljammer-monthly',
    website: 'www.example.com/spelljammer-monthly',
  },
];
