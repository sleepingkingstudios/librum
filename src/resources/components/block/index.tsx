import * as React from 'react';

import type { Response } from '@api';
import { LoadingOverlay } from '@components/loading-overlay';
import type {
  ConfiguredDataBlock,
  DataBlockData,
} from '@resources/types';
import { singularize } from '@utils/inflector';
import { titleCase } from '@utils/text';

type ResourceBlockProps = {
  Block: ConfiguredDataBlock,
  loader?: JSX.Element,
  response: Response,
  resourceName: string,
  singularName?: string,
};

const renderContent = ({
  Block,
  loader,
  resourceName,
  response,
  singularName,
}: {
  Block: ConfiguredDataBlock,
  loader?: JSX.Element,
  resourceName: string,
  response: Response,
  singularName?: string,
}): JSX.Element => {
  const {
    isLoading,
    hasData,
  } = response;

  if (isLoading) {
    return renderLoader({ loader, resourceName, singularName });
  }

  if (!hasData) {
    return (<Block data={{}} name={resourceName} />);
  }

  const data = response.data as DataBlockData;

  return (<Block data={data} name={resourceName} />);
};

const renderLoader = ({
  loader,
  resourceName,
  singularName,
}: {
  loader?: JSX.Element,
  resourceName: string,
  singularName?: string,
}): JSX.Element => {
  if (loader) { return(loader); }

  const message =
    `Loading ${titleCase(singularName || singularize(resourceName))}...`;

  return (<LoadingOverlay message={message} />);
};

export const ResourceBlock = ({
  Block,
  loader,
  response,
  resourceName,
  singularName,
}: ResourceBlockProps): JSX.Element => (
  <div className="min-h-[10rem] relative resource-block">
    { renderContent({ Block, loader, resourceName, singularName, response }) }
  </div>
);
