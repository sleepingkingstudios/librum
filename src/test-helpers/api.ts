import fetchMock from 'jest-fetch-mock';
import {
  Middleware,
  Reducer,
  configureStore,
} from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query/react';

import {
  actions,
  reducer as session,
} from '@session';
import type { User } from '@session';
import type {
  ApiParam,
  ApiResponse,
} from '@api';

type Dispatch = (arg: unknown) => unknown;

type Endpoint = { initiate: (param: unknown) => unknown };

type GetDefaultMiddleware = () => Middleware[];

type Response = {
  data?: unknown;
  error?: {
    error: string;
  },
  status: 'fulfilled' | 'rejected';
}

type Store = ReturnType<typeof configureStore>;

interface ApiService {
  endpoints: {
    [name: string]: Endpoint;
  };
  middleware: Middleware;
  reducer: Reducer;
}

interface DispatchRequest {
  api: ApiService;
  endpoint: string;
  param?: ApiParam;
  store?: Store;
}

interface ShouldMatchTheRequest {
  headers?: { [name: string]: string };
  method?: string;
  url: string;
}

interface ShouldPerformTheMutation {
  api: ApiService;
  data: ApiResponse;
  endpoint: string;
  param?: ApiParam;
  request: ShouldMatchTheRequest;
}

interface ShouldPerformTheQuery {
  api: ApiService;
  data: ApiResponse;
  endpoint: string;
  param?: ApiParam;
  request: ShouldMatchTheRequest;
}

const authenticatedRequest = (
  { request, token }: { request: ShouldMatchTheRequest, token: string }
) => {
  const { headers, method, url } = request;
  const authenticatedHeaders = {
    ...headers,
    'Authorization': `Bearer ${token}`,
  };

  return {
    headers: authenticatedHeaders,
    method,
    url,
  };
}

const createApiStore = ({ api }: { api: ApiService }): Store => {
  const middleware = (
    getDefaultMiddleware: GetDefaultMiddleware,
  ) => getDefaultMiddleware().concat(api.middleware);
  const reducer = {
    api: api.reducer,
    session,
  };
  const store = configureStore({
    middleware,
    reducer,
  });

  setupListeners(store.dispatch);

  return store;
};

const dispatchRequest = async (
  {
    api,
    endpoint,
    param,
    store,
  }: DispatchRequest
) => {
  let configuredStore = store;

  if (!store) { configuredStore = createApiStore({ api }); }

  const dispatch: Dispatch = configuredStore.dispatch;

  const { endpoints } = api;
  const query = endpoints[endpoint];

  return await dispatch(query.initiate(param)) as Response;
};

const shouldMatchTheRequest = (
  { method = 'GET', headers = {}, url }: ShouldMatchTheRequest
) => {
  const request = fetchMock.mock.calls[0][0];

  if (typeof request === 'string') { throw(`Request??? ${request}`); }

  expect(request.method).toBe(method.toUpperCase());
  expect(request.url).toBe(url);

  expect(request.headers.get('Accept')).toBe('application/json');

  Object.entries(headers).forEach(([header, value]) => {
    expect(request.headers.get(header)).toBe(value);
  });
};

// eslint-disable-next-line jest/no-export
export const shouldPerformTheMutation = (
  {
    api,
    data,
    endpoint,
    param,
    request,
  }: ShouldPerformTheMutation
) => {
  describe('should perform the mutation', () => {
    // eslint-disable-next-line jest/expect-expect
    it('should update the data', async () => {
      await dispatchRequest({ api, endpoint, param });

      shouldMatchTheRequest(request);
    });

    describe('with a successful response', () => {
      it('should fulfill the request', async () => {
        fetchMock.once(JSON.stringify(data));

        const response = await dispatchRequest({ api, endpoint });

        expect(response.data).toEqual(data);
      });
    });

    describe('with a failing response', () => {
      it('should reject the request', async () => {
        fetchMock.mockRejectOnce(new Error('Something went wrong'));

        const response = await dispatchRequest({ api, endpoint });

        expect(response.error.error).toBe('Error: Something went wrong');
      });
    });

    describe('when the user is authenticated', () => {
      const { create } = actions;
      const user: User = {
        email: 'alan.bradley@example.com',
        id: '00000000-0000-0000-0000-000000000000',
        role: 'user',
        slug: 'alan-bradley',
        username: 'Alan Bradley',
      };
      const token = '12345';

      // eslint-disable-next-line jest/expect-expect
      it('should update the data', async () => {
        const store = createApiStore({ api });
        const dispatch: Dispatch = store.dispatch;
        const action = create({ token, user });

        dispatch(action);

        await dispatchRequest({
          api,
          endpoint,
          param,
          store,
        });

        shouldMatchTheRequest(authenticatedRequest({ request, token }));
      });
    });
  });
};

// eslint-disable-next-line jest/no-export
export const shouldPerformTheQuery = (
  {
    api,
    data,
    endpoint,
    param,
    request,
  }: ShouldPerformTheQuery
) => {
  describe('should perform the query', () => {
    // eslint-disable-next-line jest/expect-expect
    it('should query the data', async () => {
      await dispatchRequest({ api, endpoint, param });

      shouldMatchTheRequest(request);
    });

    describe('with a successful response', () => {
      it('should fulfill the request', async () => {
        fetchMock.once(JSON.stringify(data));

        const response = await dispatchRequest({ api, endpoint });
        const { status } = response;

        expect(response.data).toEqual(data);
        expect(status).toBe('fulfilled');
      });
    });

    describe('with a failing response', () => {
      it('should reject the request', async () => {
        fetchMock.mockRejectOnce(new Error('Something went wrong'));

        const response = await dispatchRequest({ api, endpoint });
        const { error: { error }, status } = response;

        expect(error).toBe('Error: Something went wrong');
        expect(status).toBe('rejected');
      });
    });

    describe('when the user is authenticated', () => {
      const { create } = actions;
      const user: User = {
        email: 'alan.bradley@example.com',
        id: '00000000-0000-0000-0000-000000000000',
        role: 'user',
        slug: 'alan-bradley',
        username: 'Alan Bradley',
      };
      const token = '12345';

      // eslint-disable-next-line jest/expect-expect
      it('should query the data', async () => {
        const store = createApiStore({ api });
        const dispatch: Dispatch = store.dispatch;
        const action = create({ token, user });

        dispatch(action);

        await dispatchRequest({
          api,
          endpoint,
          param,
          store,
        });

        shouldMatchTheRequest(authenticatedRequest({ request, token }));
      });
    });
  });
};
