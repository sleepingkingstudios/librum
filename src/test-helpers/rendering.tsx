import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
