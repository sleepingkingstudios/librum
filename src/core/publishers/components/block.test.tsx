import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { PublisherBlock } from './block';
import { publishers } from '../test-helpers';

describe('<PublisherBlock />', () => {
  describe('when there is no matching data', () => {
    const data = {};

    it('should render the empty block', () => {
      const expectedTerms = ['Name', 'Slug', 'Website'];
      const expectedDfns  = [' ', ' ', ' '];
      const { getAllByRole } = render(
        <PublisherBlock data={data} />,
        { router: true },
      );

      const terms = getAllByRole('term');
      const dfns = getAllByRole('definition');

      expect(terms.map(term => term.textContent)).toEqual(expectedTerms);
      expect(dfns.map(dfn => dfn.textContent)).toEqual(expectedDfns);
    });

    it('should render the publishers link', () => {
      const { getByRole } = render(
        <PublisherBlock data={data} />,
        { router: true },
      );
      const link = getByRole('link', { name: 'Back to Publishers' });

      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', '/publishers');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <PublisherBlock data={data} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when there is publisher data', () => {
    const data = { publisher: publishers[0] };

    it('should render the publisher block', () => {
      const expectedTerms = ['Name', 'Slug', 'Website'];
      const expectedDfns = [
        'Crystal Sphere Publications',
        'crystal-sphere-publications',
        'www.example.com/crystal-sphere-publications',
      ];
      const { getAllByRole } = render(
        <PublisherBlock data={data} />,
        { router: true },
      );

      const terms = getAllByRole('term');
      const dfns = getAllByRole('definition');

      expect(terms.map(term => term.textContent)).toEqual(expectedTerms);
      expect(dfns.map(dfn => dfn.textContent)).toEqual(expectedDfns);
    });

    it('should render the website link', () => {
      const { getByRole } = render(
        <PublisherBlock data={data} />,
        { router: true },
      );
      const link = getByRole(
        'link',
        { name: 'www.example.com/crystal-sphere-publications' },
      );

      expect(link).toBeVisible();
      expect(link).toHaveAttribute(
        'href',
        'https://www.example.com/crystal-sphere-publications',
      );
    });

    it('should render the publishers link', () => {
      const { getByRole } = render(
        <PublisherBlock data={data} />,
        { router: true },
      );
      const link = getByRole('link', { name: 'Back to Publishers' });

      expect(link).toBeVisible();
      expect(link).toHaveAttribute('href', '/publishers');
    });

    it('should match the snapshot', () => {
      const { asFragment } = render(
        <PublisherBlock data={data} />,
        { router: true },
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
