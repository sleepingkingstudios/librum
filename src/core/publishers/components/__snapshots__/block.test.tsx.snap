// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublisherBlock /> when there is no matching data should match the snapshot 1`] = `
<DocumentFragment>
  <dl
    class="data-list"
  >
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Name
      </dt>
      <dd
        class="data-list-value"
      >
         
      </dd>
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Slug
      </dt>
      <dd
        class="data-list-value"
      >
         
      </dd>
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Website
      </dt>
      <dd
        class="data-list-value"
      >
         
      </dd>
    </div>
  </dl>
  <a
    href="/publishers"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-left-long mr-1"
      data-icon="left-long"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M177.5 414c-8.8 3.8-19 2-26-4.6l-144-136C2.7 268.9 0 262.6 0 256s2.7-12.9 7.5-17.4l144-136c7-6.6 17.2-8.4 26-4.6s14.5 12.5 14.5 22l0 72 288 0c17.7 0 32 14.3 32 32l0 64c0 17.7-14.3 32-32 32l-288 0 0 72c0 9.6-5.7 18.2-14.5 22z"
        fill="currentColor"
      />
    </svg>
    Back to Publishers
  </a>
</DocumentFragment>
`;

exports[`<PublisherBlock /> when there is publisher data should match the snapshot 1`] = `
<DocumentFragment>
  <dl
    class="data-list"
  >
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Name
      </dt>
      <dd
        class="data-list-value"
      >
        Crystal Sphere Publications
      </dd>
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Slug
      </dt>
      <dd
        class="data-list-value"
      >
        crystal-sphere-publications
      </dd>
    </div>
    <div
      class="data-list-item"
    >
      <dt
        class="data-list-label"
      >
        Website
      </dt>
      <dd
        class="data-list-value"
      >
        <a
          href="https://www.example.com/crystal-sphere-publications"
        >
          <span>
            www.example.com/crystal-sphere-publications
          </span>
        </a>
      </dd>
    </div>
  </dl>
  <a
    href="/publishers"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-left-long mr-1"
      data-icon="left-long"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M177.5 414c-8.8 3.8-19 2-26-4.6l-144-136C2.7 268.9 0 262.6 0 256s2.7-12.9 7.5-17.4l144-136c7-6.6 17.2-8.4 26-4.6s14.5 12.5 14.5 22l0 72 288 0c17.7 0 32 14.3 32 32l0 64c0 17.7-14.3 32-32 32l-288 0 0 72c0 9.6-5.7 18.2-14.5 22z"
        fill="currentColor"
      />
    </svg>
    Back to Publishers
  </a>
</DocumentFragment>
`;
